### PR TITLE
Reimplemented generalized plane strain using scalar kernel and update…

### LINKED ISSUE #7781: Reimplementation of generalized plane strain using scalar kernel
### Description of the enhancement or error report

Reimplemented the generalized plane strain model using scalar kernel. The previous version in solid mechanics treats the out-of-plane strain_zz as a nodal DOF and use constraint system to set them all equal. In the reimplementation in tensor mechanics, out-of-plane strain_zz was taken as scalar variable and hence scalar kernel is used to computeResidual() and computeJacobian() with the value returned from a user object.

### PR DESCRIPTION
**Pull request by hchen139**
- Reimplemented generalized plane strain using scalar kernel in tensor mechanics. Changes were made for the out-of-plane strain_zz from nonlinear variable to scalar variable.
- Changes in PR #7749 is required and included
- Modified and migrated existing test cases from solid mechanics to tensor mechanics
- Regolded the tests without volumetric correction from solid mechanics
- Closes #7781

### CODE REVIEW

**Comment by hchen139** [comment]
@bwspenc this is the PR for reimplementation of generalized plane strain model in tensor mechanics
@dschwen Would you mind to have a look at it and see whether you like it or not? :)
@andrsd changes in PR #7749 is required and included

File: framework/src/base/FEProblem.C [hunk 1]
```diff
@@ -3561,6 +3564,9 @@ FEProblem::computeJacobian(NonlinearImplicitSystem & sys, const NumericVector<Nu
     execTransfers(EXEC_NONLINEAR);
     execMultiApps(EXEC_NONLINEAR);
 
+    for (unsigned int tid = 0; tid < n_threads; tid++)
+      reinitScalars(tid);
```
Comment (dschwen) [inline framework/src/base/FEProblem.C]: Is this necessary here? We already reinitialize scalars in `NonlinearSystem::computeJacobianInternal`, `NonlinearSystem::computeResidualInternal`, and other places. Is it to get the scalars initialized for the UserObjects? It probably doesn't add a noticeable overheat, but it adds it for everyone and each residual computation, so we should be careful here.

File: framework/src/base/FEProblem.C [hunk 2]
```diff
@@ -3499,6 +3499,9 @@ FEProblem::computeResidualType(const NumericVector<Number>& soln, NumericVector<
 
   execMultiApps(EXEC_LINEAR);
 
+  for (unsigned int tid = 0; tid < n_threads; tid++)
+    reinitScalars(tid);
```
Comment (dschwen) [inline framework/src/base/FEProblem.C]: See comment below. Do we need this here?

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#ifndef GENERALIZEDPLANESTRAINOFFDIAG_H
+#define GENERALIZEDPLANESTRAINOFFDIAG_H
+
+#include "Kernel.h"
+#include "RankTwoTensor.h"
+#include "RankFourTensor.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h]: use forward declarations here and use include only in the .C file. We've worked hard to keep the include statements to a minimum to improve our compile times.

```
class RankTwoTensor;
class RankFourTensor;
```

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+template<>
+InputParameters validParams<GeneralizedPlaneStrainOffDiag>();
+
+class GeneralizedPlaneStrainOffDiag : public Kernel
+{
+public:
+  GeneralizedPlaneStrainOffDiag(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual(){return 0;}
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h]: spacing

File: modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  virtual void computeOffDiagJacobianScalar(unsigned int jvar);
+
+  virtual void computeDispOffDiagJacobianScalar(unsigned int component, unsigned int jvar);
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/GeneralizedPlaneStrainOffDiag.h]: please add Doxygen comments to document these methods

File: modules/tensor_mechanics/include/materials/ComputePlaneFiniteStrain.h [hunk 6]
```diff
@@ -22,6 +21,10 @@ class ComputePlaneFiniteStrain : public Compute2DFiniteStrain
 protected:
   virtual Real computeDeformGradZZ();
   virtual Real computeDeformGradZZold();
+
+private:
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ComputePlaneFiniteStrain.h]: why private? (applies to everything below as well)

File: modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenStrain.h [hunk 7]
```diff
@@ -8,6 +8,7 @@
 #define COMPUTETHERMALEXPANSIONEIGENSTRAIN_H
 
 #include "ComputeStressFreeStrainBase.h"
+#include "RankTwoTensor.h"
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenStrain.h]: forward declare the tensor class

File: modules/tensor_mechanics/include/scalarkernels/GeneralizedPlaneStrainDiag.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+class GeneralizedPlaneStrainDiag : public ScalarKernel
+{
+public:
+  GeneralizedPlaneStrainDiag(const InputParameters & parameters);
+
+  virtual void reinit(){};
+  virtual void computeResidual();
+  virtual void computeJacobian();
+
+private:
```
Comment (dschwen) [inline modules/tensor_mechanics/include/scalarkernels/GeneralizedPlaneStrainDiag.h]: private is really not necessary here

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  {
+    DenseMatrix<Number> & kne = _assembly.jacobianBlock(jvar, _var.number());
+    MooseVariableScalar & jv = _sys.getScalarVariable(_tid, jvar);
+
+    for (_i = 0; _i < _test.size(); _i++)
+      for (_j = 0; _j < jv.order(); _j++)
+        for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+        {
+          kne(_j, _i) += _JxW[_qp] * _coord[_qp] * _dstress_dT[_qp](2, 2) * _test[_i][_qp];
+        }
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: no braces here

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+GeneralizedPlaneStrainOffDiag::computeDispOffDiagJacobianScalar(unsigned int component, unsigned int jvar)
+{
+  if (jvar == _strain_zz_var)
+  {
+    DenseMatrix<Number> & ken = _assembly.jacobianBlock(_var.number(), jvar);
+    DenseMatrix<Number> & kne = _assembly.jacobianBlock(jvar, _var.number());
+    MooseVariableScalar & jv = _sys.getScalarVariable(_tid, jvar);
+
+    for (_i = 0; _i < _test.size(); _i++)
+      for (_j = 0; _j < jv.order(); _j++)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: everywhere in tensor mechanics we use pre increment in loops `++_i` etc.

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C [hunk 11]
```diff
@@ -25,8 +25,10 @@ ComputeStressBase::ComputeStressBase(const InputParameters & parameters) :
     _stress(declareProperty<RankTwoTensor>(_base_name + "stress")),
     _elastic_strain(declareProperty<RankTwoTensor>(_base_name + "elastic_strain")),
     _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_base_name + "elasticity_tensor")),
+    _dstrain_dT(getDefaultMaterialProperty<RankTwoTensor>(_base_name + "dstrain_dT")),
     _extra_stress(getDefaultMaterialProperty<RankTwoTensor>(_base_name + "extra_stress")),
     _Jacobian_mult(declareProperty<RankFourTensor>(_base_name + "Jacobian_mult")),
+    _dstress_dT(declareProperty<RankTwoTensor>(_base_name + "dstress_dT")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeStressBase.C]: Please use the DerivativeMaterial interface for this. We should not have any hardcoded _derivative_ property names in tensor mechanics...

...But this class doesn't even couple to the temperature variable...

...so this class does not seem like the right place to declare this property at all.

File: modules/tensor_mechanics/src/materials/ComputeThermalExpansionEigenStrain.C [hunk 12]
```diff
@@ -24,6 +24,7 @@ ComputeThermalExpansionEigenStrain::ComputeThermalExpansionEigenStrain(const Inp
     _has_incremental_strain(hasMaterialProperty<RankTwoTensor>(_base_name + "strain_increment")),
     _temperature_old(_has_incremental_strain ? & coupledValueOld("temperature") : NULL),
     _thermal_expansion_coeff(getParam<Real>("thermal_expansion_coeff")),
+    _dstrain_dT(declareProperty<RankTwoTensor>(_base_name + "dstrain_dT")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeThermalExpansionEigenStrain.C]: This class couples to temperature and should use `declareMaterialPropertyDerivative`

File: modules/tensor_mechanics/src/scalarkernels/GeneralizedPlaneStrainDiag.C [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "GeneralizedPlaneStrainDiag.h"
+#include "Assembly.h"
+
+template<>
+InputParameters validParams<GeneralizedPlaneStrainDiag>()
+{
+  InputParameters params = validParams<ScalarKernel>();
+  params.addClassDescription("Generalized Plane Strain Scalar Kernel");
+  params.addRequiredParam<UserObjectName>("gps_uo", "The name of the GeneralizedPlaneStrainUO UserObject");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/scalarkernels/GeneralizedPlaneStrainDiag.C]: I'm not a fan of this parameter name. maybe we can come up with something less cryptic and without a system suffix like `_uo`. Why not just `generalized_plane_strain`

File: modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUO.C [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+// libmesh includes
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<GeneralizedPlaneStrainUO>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  params.addClassDescription("Generalized Plane Strain UserObject to provide Residual and diagonal Jacobian entry");
+  params.addParam<FunctionName>("function", "0", "Function used to prescribe pressure in the out-of-plane direction");
+  params.addParam<Real>("factor", 1.0, "Scale factor applied to prescribed pressure");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/userobjects/GeneralizedPlaneStrainUO.C]: I'd prefer to have this named `pressure` or something that describes what the _purpose_ is rather than what kind of object is expected (we already have that information in the parameter type).

File: framework/src/base/FEProblem.C (hunk 2 above)
Comment (hchen139) [inline framework/src/base/FEProblem.C]: @andrsd Would you like to respond to this?

File: modules/tensor_mechanics/include/scalarkernels/GeneralizedPlaneStrainDiag.h [hunk 15]
```diff
@@ -0,0 +1,32 @@
+/****************************************************************/
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/scalarkernels/GeneralizedPlaneStrainDiag.h]: I think this should just be called `GeneralizedPlaneStrain` because it computes the residual corresponding to the scalar variable.

File: modules/tensor_mechanics/src/scalarkernels/GeneralizedPlaneStrainDiag.C (hunk 13 above)
Comment (bwspenc) [inline modules/tensor_mechanics/src/scalarkernels/GeneralizedPlaneStrainDiag.C]: Along those lines, the name of GeneralizedPlaneStrainUO is cryptic in the same way.  I guess I'd call it GeneralizedPlaneStrainUserObject.

File: modules/solid_mechanics/tests/generalized_plane_strain/tests [hunk 16]
```diff
@@ -1,38 +0,0 @@
-[Tests]
```
Comment (bwspenc) [inline modules/solid_mechanics/tests/generalized_plane_strain/tests]: We still need these tests in solid_mechanics, don't we?

File: framework/src/base/FEProblem.C (hunk 2 above)
Comment (andrsd) [inline framework/src/base/FEProblem.C]: Yes, we do need that, see #7749. The reason is:

>Since user objects can couple scalar variables in, we should reinit the
values of all scalar variables before we execute the user object loop.

The same has to be done in `computeJacobian`

File: modules/tensor_mechanics/tests/generalized_plane_strain/tests [hunk 17]
```diff
@@ -1,10 +1,26 @@
 [Tests]
```
Comment (jiangwen84) [inline modules/tensor_mechanics/tests/generalized_plane_strain/tests]: Since you have modified material class of small total strain, incremental strain and finite strain for Generalized Plane Strain case, I would suggest you add several tests for those formulations.

Comment (hchen139) [inline modules/tensor_mechanics/tests/generalized_plane_strain/tests]: @bwspenc should we add those tests? If so, this can be addressed in another PR. @jiangwen84

File: modules/solid_mechanics/tests/generalized_plane_strain/tests (hunk 16 above)
Comment (hchen139) [inline modules/solid_mechanics/tests/generalized_plane_strain/tests]: @bwspenc those tests are specifically for generalized plane strain model. Since we reimplemented the model and updated the tests in tensor mechanics, should we still keep them in solid mechanics?

File: modules/tensor_mechanics/tests/generalized_plane_strain/tests (hunk 17 above)
Comment (sapitts) [inline modules/tensor_mechanics/tests/generalized_plane_strain/tests]: Please do add in the tests for all three different strain formulations so that we have testing code coverage for the new strain materials.  You could likely use the same test for the small strain and the small incremental strain, using command line arguments in the test file, so that you would only need to add two new tests.

File: modules/solid_mechanics/tests/generalized_plane_strain/tests (hunk 16 above)
Comment (bwspenc) [inline modules/solid_mechanics/tests/generalized_plane_strain/tests]: As we implement things in tensor_mechanics, we are still keeping all of the tests in solid_mechanics. We generally want to keep all tests around until we get rid of the code. The solid_mechanics code is being widely used, and will be for some time, so we want to make sure it stays working.

Comment (hchen139) [inline modules/solid_mechanics/tests/generalized_plane_strain/tests]: I have brought back those tests in solid mechanics

File: modules/tensor_mechanics/tests/generalized_plane_strain/tests (hunk 17 above)
Comment (hchen139) [inline modules/tensor_mechanics/tests/generalized_plane_strain/tests]: Will do this in a separate PR

File: modules/combined/tests/generalized_plane_strain_tm_contact/generalized_plane_strain_tm_contact_field.i [hunk 18]
```diff
@@ -94,9 +94,9 @@
 
 [ScalarKernels]
   [./gps]
-    type = GeneralizedPlaneStrainDiag
+    type = GeneralizedPlaneStrain
     variable = strain_zz
-    gps_uo = gpsuo
+    generalized_plane_strain_userobject = gpsuo
```
Comment (dschwen) [inline modules/combined/tests/generalized_plane_strain_tm_contact/generalized_plane_strain_tm_contact_field.i]: We are moving away from naming parameters `..._userobject`. It is already encoded in the parameter type that a _userobject_ name is expected.

File: modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C [hunk 19]
```diff
@@ -27,10 +29,10 @@ InputParameters validParams<GeneralizedPlaneStrainOffDiag>()
 }
 
 GeneralizedPlaneStrainOffDiag::GeneralizedPlaneStrainOffDiag(const InputParameters & parameters) :
-   Kernel(parameters),
+   DerivativeMaterialInterface<Kernel>(parameters),
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    _Jacobian_mult(getMaterialProperty<RankFourTensor>(_base_name + "Jacobian_mult")),
-   _dstress_dT(getMaterialProperty<RankTwoTensor>(_base_name + "dstress_dT")),
+   _thermal_expansion_tensor(getDefaultMaterialProperty<RankTwoTensor>(_base_name + "thermal_expansion_tensor")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C]: `+ "_thermal_expansion_tensor"` would produce slightly nicer property names.

**Comment by dschwen** [comment]
Looks like all comments were addressed.

**Review by dschwen — APPROVED** [review]
